File: Python with BDD Cucumber/tests/pages/login_page.py
from selenium.webdriver.common.by import By
from tests.utils.base_page import BasePage
from tests.utils.base_element import BaseElement


class LoginPage(BasePage):
    """Login page objects for ParaBank"""
    
    # Locators
    USERNAME_INPUT = (By.NAME, "username")
    PASSWORD_INPUT = (By.NAME, "password")
    LOGIN_BUTTON = (By.CSS_SELECTOR, "input[value='Log In']")
    FORGOT_LOGIN_LINK = (By.LINK_TEXT, "Forgot login info?")
    REGISTER_LINK = (By.LINK_TEXT, "Register")
    ERROR_MESSAGE = (By.CSS_SELECTOR, ".error")
    
    def __init__(self, driver):
        super().__init__(driver)
        self.base_url = "https://parabank.parasoft.com/"
    
    def open_login_page(self):
        """Open ParaBank login page"""
        self.navigate_to(self.base_url)
        self.logger.info("Opened ParaBank login page")
    
    def enter_username(self, username):
        """Enter username"""
        element = BaseElement(self.driver, self.USERNAME_INPUT)
        element.send_keys(username)
    
    def enter_password(self, password):
        """Enter password"""
        element = BaseElement(self.driver, self.PASSWORD_INPUT)
        element.send_keys(password)
    
    def click_login(self):
        """Click login button"""
        element = BaseElement(self.driver, self.LOGIN_BUTTON)
        element.click()
    
    def login(self, username, password):
        """Complete login process"""
        self.enter_username(username)
        self.enter_password(password)
        self.click_login()
        self.logger.info(f"Attempted login with username: {username}")
    
    def is_login_page_displayed(self):
        """Check if login page is displayed"""
        element = BaseElement(self.driver, self.USERNAME_INPUT)
        return element.is_displayed()
    
    def get_error_message(self):
        """Get error message"""
        element = BaseElement(self.driver, self.ERROR_MESSAGE)
        return element.get_text()
    
    def click_register_link(self):
        """Click register link"""
        element = BaseElement(self.driver, self.REGISTER_LINK)
        element.click()
    
    def click_forgot_login_link(self):
        """Click forgot login link"""
        element = BaseElement(self.driver, self.FORGOT_LOGIN_LINK)
        element.click()

File: Java with BDD Cucumber/src/main/java/com/parabank/pages/LoginPage.java
package com.parabank.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import com.parabank.base.BaseClass;

/**
 * Login Page Object Model
 */
public class LoginPage extends BaseClass {
    // Locators
    private By usernameField = By.name("customer.username");
    private By passwordField = By.name("customer.password");
    private By loginButton = By.xpath("//input[@value='Log In']");
    private By errorMessage = By.cssSelector("p.error");
    private By welcomeMessage = By.cssSelector("h2");

    public LoginPage(WebDriver driver) {
        this.setDriver(driver);
    }

    /**
     * Navigate to login page
     */
    public void navigateToLoginPage() {
        navigateToUrl("https://www.parabank.example.com/parabank/index.htm");
    }

    /**
     * Enter username
     *
     * @param username Username to enter
     */
    public void enterUsername(String username) {
        typeText(usernameField, username);
    }

    /**
     * Enter password
     *
     * @param password Password to enter
     */
    public void enterPassword(String password) {
        typeText(passwordField, password);
    }

    /**
     * Click login button
     */
    public void clickLoginButton() {
        clickElement(loginButton);
    }

    /**
     * Get error message
     *
     * @return Error message text
     */
    public String getErrorMessage() {
        return getText(errorMessage);
    }

    /**
     * Check if error message is displayed
     *
     * @return true if error is displayed
     */
    public boolean isErrorMessageDisplayed() {
        return isElementDisplayed(errorMessage);
    }

    /**
     * Get welcome message
     *
     * @return Welcome message text
     */
    public String getWelcomeMessage() {
        return getText(welcomeMessage);
    }

    /**
     * Perform login
     *
     * @param username Username
     * @param password Password
     */
    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLoginButton();
    }
}
